### DISCUSSION TITLE
[Bug]: Using Cosmos DB: E11000 duplicate key error when new user attempts login or registration

### DISCUSSION THREAD

**Discussion by robsaccone** [danny-avila/LibreChat#1778] [Troubleshooting]
### What happened?

Using Azure Cosmos DB as backend, which (poorly) emulates Mongo DB, and as a result have encountered a number of issues.  This one is due to unique indexes being created in the users collection for each of the "social login" options (facebookId, etc.).

Apparently Cosmos DB doesn't support sparse unique indexes, i.e. treats multiple NULL values as duplicates.  I am using OpenID only, so manually deleting the other indexes in the collection solved the issue. 

Not looking for a fix here per se, as I'm beginning to realize that Cosmos DB should be consider a semi-supported choice given it's API emulation.  Adding detail for others who might try the same, and perhaps for those who might have ideas to better support both Cosmos and Mongo.

### Steps to Reproduce

1. Deploy using Azure Cosmos DB as database vs. Mongo native.
2. Have more than one user attempt login or registration
3. Get duplicate key error.

### Relevant log output

```shell
E11000 duplicate key error collection: test.users. Failed _id or unique index constraint.
```

### Screenshots

![image](https://github.com/danny-avila/LibreChat/assets/36059743/e5539878-a63a-4fc9-9432-b3ec84e3d89d)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Yeah this is a known issue for me. I might be able to remove the sparse field as I’m not sure it’s needed. Still need to investigate the effects of removing it.

I’ve heard DocumentDB is a good enterprise level, cloud alternative (AWS)

**Comment by robsaccone** [comment]
Thanks Danny. Thinking of dropping Cosmos and either containerizing Mongo as prescribed or using Atlas.  Most of my clients are big Azure shops, hence my first attempt at an Azure-first approach.  Painful so far but it's working.  About to deploy the assistants branch to Azure...

**Reply by ndrsfel** [reply]
I only had issues with CosmosDB but wanted to stay within Azure for the DB. I then decided to deploy a MongoDB Container on Azure Container Instance which has been working like a charm. If that is of general interest I can extend the Azure documentation of LibreChat✌️

**Reply by strayer** [reply]
I'd be interested in how you deployed MongoDB with ACI. I intend to deploy a demo instance of LibreChat in our Azure tenant as well. Do you mount persistent storage for MongoDB with Azure Files? I'd assume that this might cause issues with MongoDB as databases usually don't really like stuff like that.

**Reply by danny-avila** [reply]
>  I can extend the Azure documentation of LibreChat

Please do! Also for what it's worth, this PR could potentially solve the root problem here: https://github.com/danny-avila/LibreChat/pull/1814

Merging it tomorrow.

**Reply by ndrsfel** [reply]
Sorry for my slow response. I used [this](https://jussiroine.com/2019/02/an-adventure-in-containers-and-command-line-tools-running-mongodb-in-azure/) tutorial by Jussi Roine using Azure Container Instances.

@strayer you are right, the tutorial involves mounting a Blob Storage for persistent storage. This setup has been working just fine for a couple of days now with dozens of users on `v0.6.10`.

**Reply by strayer** [reply]
@ndrsfel Interesting. I wonder why that works with ACI and not with Azure Web Apps (which I tried and failed as expected due to missing flock functionality). I'll try creating the MongoDB with ACI as well… MongoDB shouldn't even manage to start if it can't get a file lock. Which MongoDB version are you using?

**Reply by strayer** [reply]
Huh. Well, I don't really understand what is the cause here but apparently not using /data/db as the mount path for the azure files share but /data/mongoaz and then overriding the container command to `mongod --dbpath=/data/mongoaz` lets the container boot up fine, at least with Azure Container Instances. Didn't try a web app again.

I had trouble getting the container to start when using the default /data/db path, just like with the webapp, but wasn't able to read out the container output log so I can't say for sure it was the same issue. I'm honestly pretty confused why this works and the web app before didn't.

All in all running MongoDB on ACI instead of using CosmosDB seems to be a viable solution to run LibreChat on Azure.

Just in case that blog article goes down, this is the command line I used to create a MongoDB instance with ACI:

```sh
az container create \
  --resource-group resource-group-name \
  --name container-name \
  --image mongo:7.0.4 \
  --dns-name-label container-dns-name \
  --ports 27017 \
  --azure-file-volume-account-name storageaccountname \
  --azure-file-volume-account-key foobar \
  --azure-file-volume-share-name filesharename \
  --azure-file-volume-mount-path /data/mongoaz \
  --command-line "mongod --dbpath=/data/mongoaz --bind_ip_all"
```

Notes:

- this expects a storage account with a files share already existing
- **this will create an unprotected public MongoDB instance - do not use this in production without changes**
- I can't say for sure how big the performance impact of running a MongoDB from Azure Files is

**Reply by ndrsfel** [reply]
Great write-up and glad it’s working for you as well. We can think of adding it to the LC documentation but I’ll personally switch to CosmosDB once the bug above is resolved.

**Reply by strayer** [reply]
It definitely makes much more sense to use a service like CosmosDB, but I'm a bit wary about potential issues in the future. CosmosDB didn't really inspire confidence in me during this endeavor.

**Reply by illgitthat** [reply]
> > I can extend the Azure documentation of LibreChat
> 
> Please do! Also for what it's worth, this PR could potentially solve the root problem here: #1814
> 
> Merging it tomorrow.

Is there any update on this PR / fix to this issue?

**Reply by danny-avila** [reply]
PR does not actually fix the issue, except presents another compatibility issue with Azure CosmosDB, that's why it hasn't been merged

**Reply by philipbrembeck** [reply]
Any updates on this?

**Reply by gijs-voogd** [reply]
anyone who has a working solution for the azure cosmos db? running it in the container works but I prefer using the paas db

**Reply by rubentalstra** [reply]
@gijs-voogd this one works: https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/

don't use the `RU`

**Comment by gijs-voogd** [comment]
thanks @rubentalstra. switched to vcore but this gives me:

![image](https://github.com/user-attachments/assets/346f7c6b-be19-4e1a-8cc1-36cf4da840aa)

![image](https://github.com/user-attachments/assets/1538bcea-54aa-4ac6-89f0-8ccccc5ffb70)

while there is nothing in the db:

![image](https://github.com/user-attachments/assets/11cb0a14-2e50-4daa-943f-aa084f7aeed4)

are you using the vcore model as well?

**Reply by rubentalstra** [reply]
very strange because it works for me

**Reply by rubenaster** [reply]
@rubentalstra I'm not sure that really RU is the problem. What MongoDB version did you configure for your CosmosDB?

I have the same issue:
```
2025-03-24T21:08:58.984295886Z 2025-03-24 21:08:58 [32minfo[39m: [32m[openidStrategy] user not found with openidId: dDUXXXXXXPGs[39m
2025-03-24T21:08:59.011230544Z 2025-03-24 21:08:59 [32minfo[39m: [32m[openidStrategy] user not found with email: user.name@my.company for openidId: dDUXXXXXXPGs[39m
2025-03-24T21:08:59.039710762Z 2025-03-24 21:08:59 [31merror[39m: [31m[openidStrategy] login failed E11000 duplicate key error collection: test.users. Failed _id or unique index constraint.[39m
2025-03-24T21:08:59.048601278Z MongoServerError: E11000 duplicate key error collection: test.users. Failed _id or unique index constraint.
2025-03-24T21:08:59.048629258Z     at InsertOneOperation.execute (/app/node_modules/mongodb/lib/operations/insert.js:51:19)
2025-03-24T21:08:59.048632996Z     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2025-03-24T21:08:59.048635949Z     at async tryOperation (/app/node_modules/mongodb/lib/operations/execute_operation.js:203:20)
2025-03-24T21:08:59.048638727Z     at async executeOperation (/app/node_modules/mongodb/lib/operations/execute_operation.js:73:16)
2025-03-24T21:08:59.048657468Z     at async Collection.insertOne (/app/node_modules/mongodb/lib/collection.js:157:16)
```

**Reply by rubentalstra** [reply]
> @rubentalstra I'm not sure that really RU is the problem. What MongoDB version did you configure for your CosmosDB?
> 
> 
> 
> I have the same issue:
> 
> ```
> 
> 2025-03-24T21:08:58.984295886Z 2025-03-24 21:08:58 [32minfo[39m: [32m[openidStrategy] user not found with openidId: dDUXXXXXXPGs[39m
> 
> 2025-03-24T21:08:59.011230544Z 2025-03-24 21:08:59 [32minfo[39m: [32m[openidStrategy] user not found with email: user.name@my.company for openidId: dDUXXXXXXPGs[39m
> 
> 2025-03-24T21:08:59.039710762Z 2025-03-24 21:08:59 [31merror[39m: [31m[openidStrategy] login failed E11000 duplicate key error collection: test.users. Failed _id or unique index constraint.[39m
> 
> 2025-03-24T21:08:59.048601278Z MongoServerError: E11000 duplicate key error collection: test.users. Failed _id or unique index constraint.
> 
> 2025-03-24T21:08:59.048629258Z     at InsertOneOperation.execute (/app/node_modules/mongodb/lib/operations/insert.js:51:19)
> 
> 2025-03-24T21:08:59.048632996Z     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> 
> 2025-03-24T21:08:59.048635949Z     at async tryOperation (/app/node_modules/mongodb/lib/operations/execute_operation.js:203:20)
> 
> 2025-03-24T21:08:59.048638727Z     at async executeOperation (/app/node_modules/mongodb/lib/operations/execute_operation.js:73:16)
> 
> 2025-03-24T21:08:59.048657468Z     at async Collection.insertOne (/app/node_modules/mongodb/lib/collection.js:157:16)
> 
> ```

@rubenaster i have it working on my azure account, with mongoDB version `5.0` or `6.0`

**Reply by rubenaster** [reply]
@rubentalstra okay, that one's on you. It really works with vCore. Thanks mate!

**Comment by rubentalstra** [comment]
@gijs-voogd https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/quickstart-portal

**Reply by gijs-voogd** [reply]
yes the vcore model works for me too. the error i had yesterday with the vcore model had to do with the database. you also get that error when the connectionstring or somethings else is wrong, not sure what did the trick. deployed a new one today with mongoDB 6.0 instead of 7.0

**Comment by oxch** [comment]
Dropping indices on `users` collection resolves the issue - drop those BEFORE any data is placed into `users` collection - remove data if already present, otherwise you won't be able to drop indices (in my experience anyways). Only leave indices for `_id`, `email` fields only and RU-type Cosmos for Mongo DB deployment should work too without this error.

**Comment by jcbartle** [comment]
We have LibreChat deployments in both Azure Commercial and Azure Government, using Cosmos DB for MongoDB in both. In Commercial, we're using vCore. However, vCore isnt' yet available in Azure Government, so we're stuck with the RU version (for now). Both instances are using Entra ID for authentication.

In our Government deploy, the user experience was getting the following response upon trying to log in:

`{"messages":"An document with that unknown already exists.","fields":"unknown"}`

The following message was in the app logs:

`[openidStrategy] login failed E11000 duplicate key error collection: test.users. Failed _id or unique index constraint.`

I can confirm that the guidance in this post does address the issue described. Here were the steps I had to take to get things working:

1. Delete the initial user from the _users_ collection.
2. Reindex the _users_ collection by running `db.users.reIndex()`
3. Drop the indexes for the unused IDPs:  
    `db.users.dropIndex("googleId_1")`  
    `db.users.dropIndex("facebookId_1")`  
    `db.users.dropIndex("samlId_1")`  
    `db.users.dropIndex("ldapId_1")`  
    `db.users.dropIndex("githubId_1")`  
    `db.users.dropIndex("discordId_1")`  
    `db.users.dropIndex("appleId_1")`

Thank you to those above who contributed to figuring this out!
